core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -0,0 +1,114 @@
+import org.infinispan.context.InvocationContextFactory;
+import org.infinispan.factories.KnownComponentNames;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.InterceptorChain;
+
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+@ThreadSafe
+public class ClusterExpirationManager<K, V> extends ExpirationManagerImpl<K, V> {
+
+   private ExecutorService asyncExecutor;
+   private AdvancedCache<K, V> cache;
+
+   @Inject
+   public void inject(AdvancedCache<K, V> cache,
+           @ComponentName(KnownComponentNames.ASYNC_OPERATIONS_EXECUTOR) ExecutorService asyncExecutor) {
+      this.cache = cache;
+      this.asyncExecutor = asyncExecutor;
+   }
+
+   @Override
+   public void processExpiration() {
+      long start = 0;
+      if (!Thread.currentThread().isInterrupted()) {
+         try {
+            if (trace) {
+               log.trace("Purging data container of expired entries");
+               start = timeService.time();
+            }
+            long currentTimeMillis = timeService.wallClockTime();
+            for (Iterator<InternalCacheEntry<K, V>> purgeCandidates = dataContainer.iteratorIncludingExpired();
+                 purgeCandidates.hasNext();) {
+               InternalCacheEntry<K, V> e = purgeCandidates.next();
+               if (e.canExpire()) {
+                  if (ExpiryHelper.isExpiredMortal(e.getLifespan(), e.getCreated(), currentTimeMillis)) {
+                     handleLifespanExpireEntry(e);
+                  } else if (ExpiryHelper.isExpiredTransient(e.getMaxIdle(), e.getLastUsed(), currentTimeMillis)) {
+                     super.handleInMemoryExpiration(e);
+                  }
+               }
+            }
+            if (trace) {
+               log.tracef("Purging data container completed in %s",
+                       Util.prettyPrintTime(timeService.timeDuration(start, TimeUnit.MILLISECONDS)));
+            }
+         } catch (Exception e) {
+            log.exceptionPurgingDataContainer(e);
+         }
+      }
+
+      if (!Thread.currentThread().isInterrupted()) {
+         persistenceManager.purgeExpired();
+      }
+   }
+
+   void handleLifespanExpireEntry(InternalCacheEntry<K, V> entry) {
+      K key = entry.getKey();
+      // We remove the entry if it is the same from the datacontainer while also submitting an async remove expired
+      // to other members of the cluster
+      // The most used case will be a miss so no extra read before
+      if (expiring.putIfAbsent(key, key) == null) {
+         long lifespan = entry.getLifespan();
+         if (trace) {
+            log.tracef("Submitting expiration removal for key %s which had lifespan of %s", key, lifespan);
+         }
+         asyncExecutor.submit(() -> {
```

Comment:
The `expiring.remove(key);` could have been executed as a continuation of the CompletableFuture, via thenAccept(), thenConsume or similar.